src/repr/src/adt/timestamp.rs

```diff
@@ -494,6 +509,110 @@ impl<T: TimestampLike> CheckedTimestamp<T> {
+            let mut nanos =
+                i64::cast_from(a.nanosecond()).checked_sub(i64::cast_from(b.nanosecond()))?;
+            let mut seconds = i64::cast_from(a.second()).checked_sub(i64::cast_from(b.second()))?;
+            let mut minutes = i64::cast_from(a.minute()).checked_sub(i64::cast_from(b.minute()))?;
+            let mut hours = i64::cast_from(a.hour()).checked_sub(i64::cast_from(b.hour()))?;
+            let mut days = i64::cast_from(a.day()).checked_sub(i64::cast_from(b.day()))?;
+            let mut months = i64::cast_from(a.month()).checked_sub(i64::cast_from(b.month()))?;
+            let mut years = i64::cast_from(a.year()).checked_sub(i64::cast_from(b.year()))?;
+
+            // Flip sign if necessary.
+            if a < b {
+                nanos = -nanos;
+                seconds = -seconds;
+                minutes = -minutes;
+                hours = -hours;
+                days = -days;
+                months = -months;
+                years = -years;
+            }
+
+            // Carry negative fields into the next higher field.
+            while nanos < 0 {
+                nanos = nanos.checked_add(NANOSECONDS_PER_SECOND)?;
+                seconds = seconds.checked_sub(1)?;
+            }
+            while seconds < 0 {
+                seconds = seconds.checked_add(SECONDS_PER_MINUTE)?;
+                minutes = minutes.checked_sub(1)?;
+            }
+            while minutes < 0 {
+                minutes = minutes.checked_add(MINUTES_PER_HOUR)?;
+                hours = hours.checked_sub(1)?;
+            }
+            while hours < 0 {
+                hours = hours.checked_add(HOURS_PER_DAY)?;
+                days = days.checked_sub(1)?
+            }
+            while days < 0 {
+                if a < b {
+                    days = num_days_in_month(a).and_then(|x| days.checked_add(x))?;
+                } else {
+                    days = num_days_in_month(b).and_then(|x| days.checked_add(x))?;
+                }
+                months = months.checked_sub(1)?;
+            }
+            while months < 0 {
+                months = months.checked_add(MONTHS_PER_YEAR)?;
+                years = years.checked_sub(1)?;
+            }
+
+            // Revert the sign back, if we flipped it originally.
+            if a < b {
+                nanos = -nanos;
+                seconds = -seconds;
+                minutes = -minutes;
+                hours = -hours;
+                days = -days;
+                months = -months;
+                years = -years;
+            }
+
+            let months = i32::try_from(years * MONTHS_PER_YEAR + months).ok()?;
+            let days = i32::try_from(days).ok()?;
+            let micros = Duration::nanoseconds(
+                nanos
+                    + (seconds * NANOSECONDS_PER_SECOND)
+                    + (minutes * NANOSECONDS_PER_MINUTE)
+                    + (hours * NANOSECONDS_PER_HOUR),
+            )
```

Comment:
Can you add brief comments as to why the unchecked arithmetic is safe.

Reply:
I updated this to use checked arithmetic